File: src/main/julia/AoC2017_03.jl
#! /usr/bin/env julia
#

include("aoc_main.jl")
if abspath(PROGRAM_FILE) == @__FILE__
    include("aocd.jl")
    using .Aocd
end

module AoC2017_03

include("aoc.jl")

struct HeadingsAndPeriods
    periods::Vector{Int}
    headings::Vector{Tuple{Int,Int}}

    HeadingsAndPeriods() = new([1, 1, 2, 2],
                               [Heading_E, Heading_N, Heading_W, Heading_S])
end

function getHeadingAndPeriod(headingsAndPeriods::HeadingsAndPeriods, t::Int)
    idx = t % 4 + 1
    period = headingsAndPeriods.periods[idx]
    headingsAndPeriods.periods[idx] = period + 2
    heading = headingsAndPeriods.headings[idx]
    return heading, period
end

mutable struct Coordinate
    x::Int
    y::Int
    j::Int
    k::Int
    headingsAndPeriods::HeadingsAndPeriods
    heading
    period::Int
end

Coordinate() = begin
    c = Coordinate(0, 0, 0, 0, HeadingsAndPeriods(), nothing, 0)
    c.heading, c.period = getHeadingAndPeriod(c.headingsAndPeriods, 0)
    return c
end

function next(c::Coordinate)
    if c.j == c.period
        c.k += 1
        c.heading, c.period = getHeadingAndPeriod(c.headingsAndPeriods, c.k)
        c.j = 0
    end
    c.x += c.heading[1]
    c.y += c.heading[2]
    c.j += 1
    return c
end

function _parse(input)
    @assert length(input) == 1
    return parse.(Int, input[1])
end

function part1(input)
    square::Int = _parse(input)
    square == 1 && return 0

    i::Int = 1
    c::Coordinate = Coordinate()
    while i < square
        i += 1
        c = next(c)
        i == square && return abs(c.x) + abs(c.y)
    end
    @assert false
end

function part2(input)
    square::Int = _parse(input)
    square == 1 && return 1

    squares::Dict{Tuple{Int,Int},Int} = Dict((0, 0) =>  1)
    c::Coordinate = Coordinate()
    while true
        c = next(c)
        value::Int = 0
        for (Δx, Δy) in OCTANTS
            neighbour = (c.x + Δx, c.y + Δy)
            value += haskey(squares, neighbour) ? squares[neighbour] : 0
        end
        squares[(c.x, c.y)] = value
        value > square && return value
    end
end

function samples()
    @assert part1(("1",)) == 0
    @assert part1(("12",)) == 3
    @assert part1(("23",)) == 2
    @assert part1(("1024",)) == 31
    @assert part2(("1",)) == 1
    @assert part2(("2",)) == 4
    @assert part2(("3",)) == 4
    @assert part2(("4",)) == 5
    @assert part2(("5",)) == 10
end

end # module AoC2017_03

aoc_main(@__FILE__, ARGS, 2017, 3)

File: src/main/python/AoC2017_03.py
#! /usr/bin/env python3
#
# Advent of Code 2017 Day 3
#
from collections.abc import Generator
from aoc import my_aocd
from aoc.geometry import Position, Direction
import aocd


class DirectionsAndPeriods:
    def __init__(self):
        self.periods = [1, 1, 2, 2]
        self.directions = [
            Direction.RIGHT,
            Direction.UP,
            Direction.LEFT,
            Direction.DOWN,
        ]

    def apply(self, t: int) -> tuple[Direction, int]:
        idx = t % 4
        period = self.periods[idx]
        self.periods[idx] = period + 2
        direction = self.directions[idx]
        return direction, period


def _parse(inputs: tuple[str]) -> int:
    assert len(inputs) == 1
    return int(inputs[0])


def _coordinates() -> Generator[tuple[int, int]]:
    x = y = j = k = 0
    directions_and_periods = DirectionsAndPeriods()
    direction, period = directions_and_periods.apply(k)

    while True:
        if j == period:
            k += 1
            direction, period = directions_and_periods.apply(k)
            j = 0
        x += direction.x
        y += direction.y
        j += 1
        yield (x, y)


def part_1(inputs: tuple[str]) -> int:
    square = _parse(inputs)
    if square == 1:
        return 0
    coords = _coordinates()
    i = 1
    while i < square:
        i += 1
        x, y = next(coords)
        if i == square:
            return Position.of(x, y).manhattan_distance_to_origin()
    raise ValueError


def part_2(inputs: tuple[str]) -> int:
    square = _parse(inputs)
    if square == 1:
        return 1
    squares = dict()
    squares[(0, 0)] = 1
    coords = _coordinates()
    while True:
        x, y = next(coords)
        value = 0
        for d in Direction.octants():
            neighbour = (x + d.x, y + d.y)
            value += squares[neighbour] if neighbour in squares else 0
        squares[(x, y)] = value
        if value > square:
            return value


def main() -> None:
    puzzle = aocd.models.Puzzle(2017, 3)
    my_aocd.print_header(puzzle.year, puzzle.day)

    assert part_1(("1",)) == 0
    assert part_1(("12",)) == 3
    assert part_1(("23",)) == 2
    assert part_1(("1024",)) == 31
    assert part_2(("1",)) == 1
    assert part_2(("2",)) == 4
    assert part_2(("3",)) == 4
    assert part_2(("4",)) == 5
    assert part_2(("5",)) == 10

    inputs = my_aocd.get_input(2017, 3, 1)
    result1 = part_1(inputs)
    print(f"Part 1: {result1}")
    result2 = part_2(inputs)
    print(f"Part 2: {result2}")
    my_aocd.check_results(puzzle, result1, result2)


if __name__ == "__main__":
    main()
